### PR TITLE
Bayesian optimization workflow

### PR DESCRIPTION
**Pull request by nikhilwbhatia**
Workflow is in bayes.py, uses basic MNIST model in mnist folder. Would appreciate any comments on the code 👍!

Haven't spent much time making this easy for users to run, but if you want to run it cd into main bayesopt-workflow folder and use this command:

spell workflow --repo label=. --pip bayesian-optimization -- python bayes.py

### CODE REVIEW

**Comment by nikhilwbhatia** [comment]
screen shot 2019-01-17 at 5 14 19 pm

Sample output with 3 iterations of 3 parallel runs each

File: bayesopt-workflow/mnist/.floydignore [hunk 1]
```diff
@@ -0,0 +1,14 @@
+
```
Comment (mariusbld) [inline bayesopt-workflow/mnist/.floydignore]: Floyd?! 🤔

File: bayesopt-workflow/cifar.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
+'''Train a simple deep CNN on the CIFAR10 small images dataset.
```
Comment (bgardiner) [inline bayesopt-workflow/cifar.py]: was this code taken from somewhere else? If so, we should probably put an attribution if we are going to put it on our public repo

File: bayesopt-workflow/bayes.py [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+        #TODO validate at args parsing level
+        raise Exception('Invalid metric type ({0})'.format(args.type))
+
+    r.wait_status(client.runs.COMPLETE)
+
+    print("\t[{0}] run output: {1} ({2})".format(r.id, round(metric_value, 2), args.metric))
+
+    return metric_value if args.optimize == 'maximize' else (-1)*metric_value
+
+def f2(x):
```
[the remark this answers is not in the document]
Comment (bgardiner) [reply bayesopt-workflow/bayes.py]: is this function used? If not, maybe delete

File: bayesopt-workflow/bayes.py [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+
+    def finish(self, optimizer):
+        optimizer.register(self.last_param, self.last_output)
+
+def black_box_function(**pvals):
+    """
+    black_box_function required by Bayesian Optimizer
+
+    @param pvals, a flexible length dictionary mapping hyperparameters
+    to chosen values
```
Comment (bgardiner) [inline bayesopt-workflow/bayes.py]: could you clarify what keys and values of dictionary should be?

File: bayesopt-workflow/bayes.py [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+parser.add_argument('--metric', type=str,
+                    dest='metric', help='the metric by which your model will be evaluated',
+                    default="test_acc")
+parser.add_argument('--parallel-runs', type=int,
+                    dest='parallel', help='number of runs that will be spawned in parallel',
+                    default=3)
+parser.add_argument('--iterations', type=int,
+                    dest='iterations', help='number of times new parallel runs will be spawned',
+                    default=3)
+parser.add_argument('--optimize', type=str,
```
[the remark this answers is not in the document]
Comment (nikhilwbhatia) [reply bayesopt-workflow/bayes.py]: What if users want to maximize the avg validation accuracy vs minimize the avg validation loss?

File: bayesopt-workflow/bayes.py [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+    # get metric value
+    metrics = []
+    for m in r.metrics(metric_name=args.metric, follow=True):
+        metrics.append(m[2])
+    if not metrics:
+        raise Exception('Run ended prematurely or metric ({0}) was not found (check run logs)'.format(args.metric))
+
+    if args.type == 'average':
+        metric_value = reduce(lambda x, y: x + y, metrics) / len(metrics)
```
[the remark this answers is not in the document]
Comment (nikhilwbhatia) [reply bayesopt-workflow/bayes.py]: why average specifically? added that^ but could also cast metric_value to float instead

File: bayesopt-workflow/bayes.py [hunk 7]
```diff
@@ -0,0 +1,180 @@
...
+    return param_to_min_max
+
+#set up optimizer
+pbounds = parse_params()
+# print('pbounds are -> {0}'.format(pbounds))
+optimizer = BayesianOptimization(
+    f=black_box_function,
+    pbounds=pbounds,
+    verbose=2,
+    random_state=1
```
[the remark this answers is not in the document]
Comment (nikhilwbhatia) [reply bayesopt-workflow/bayes.py]: not sure what you mean by this^

File: bayesopt-workflow/cifar.py [hunk 8]
```diff
@@ -0,0 +1,145 @@
...
+        featurewise_std_normalization=False,  # divide inputs by std of the dataset
+        samplewise_std_normalization=False,  # divide each input by its std
+        zca_whitening=False,  # apply ZCA whitening
+        zca_epsilon=1e-06,  # epsilon for ZCA whitening
+        rotation_range=0,  # randomly rotate images in the range (degrees, 0 to 180)
+        # randomly shift images horizontally (fraction of total width)
+        width_shift_range=0.1,
+        # randomly shift images vertically (fraction of total height)
+        height_shift_range=0.1,
+        shear_range=0.,  # set range for random shear
```
Comment (ianvonseggern1) [inline bayesopt-workflow/cifar.py]: nit should these all be 0.0?
